cloudnetwork/v1/types.go

```diff
@@ -0,0 +1,101 @@
+type CloudPrivateIPConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// spec is the definition of the desired private IP request(s).
+	// +kubebuilder:validation:Required
+	// +required
+	Spec CloudPrivateIPConfigSpec `json:"spec" protobuf:"bytes,2,opt,name=spec,casttype=CloudPrivateIPConfigSpec"`
+	// status is the observed status of the desired private IP request(s). Read-only.
+	// +kubebuilder:validation:Optional
+	// +optional
+	Status CloudPrivateIPConfigStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status,casttype=CloudPrivateIPConfigStatus"`
+}
+
+// CloudPrivateIPConfigSpec specifies the requested list of private IPs and their corresponding node assignment.
+// +k8s:openapi-gen=true
+type CloudPrivateIPConfigSpec struct {
+	// +kubebuilder:validation:Required
+	// +required
+	Items []CloudPrivateIPConfigSpecItem `json:"items" protobuf:"bytes,1,opt,name=items"`
+}
+
+// CloudPrivateIPConfigStatus specifies the requested list of private IPs and their corresponding node assignment,
+// together with their assignment conditions.
+// +k8s:openapi-gen=true
+type CloudPrivateIPConfigStatus struct {
+	// +kubebuilder:validation:Required
+	// +required
+	Items []CloudPrivateIPConfigCondition `json:"items" protobuf:"bytes,1,opt,name=items"`
+}
+
+// CloudPrivateIPConfigSpecItem is a tuple consisting of a node name and IP address.
+// +k8s:openapi-gen=true
+type CloudPrivateIPConfigSpecItem struct {
+	// node is the node name, as specified by the Kubernetes field: `node.metadata.name`
+	// +kubebuilder:validation:Required
+	// +required
+	Node string `json:"node" protobuf:"bytes,1,opt,name=node"`
+	// ip is the specified IP address, can be IPv4 or IPv6
+	// +kubebuilder:validation:Required
+	// +required
+	IP string `json:"ip" protobuf:"bytes,2,opt,name=ip"`
+}
+
+// ConditionStatus specifies the current condition status of the CloudPrivateIPConfigCondition
+type ConditionStatus string
+
+const (
+	// Error defines a failed cloud assignment of the requested IP address.
+	Error ConditionStatus = "Error"
+
+	// Pending defines a pending cloud assignment of the requested IP address.
+	Pending ConditionStatus = "Pending"
+
+	// Success defines a successful cloud assignment of the requested IP address.
+	Success ConditionStatus = "Success"
+)
+
+// CloudPrivateIPConfigCondition is the assignment condition associated with each requested CloudPrivateIPConfigSpecItem
+// +k8s:openapi-gen=true
+type CloudPrivateIPConfigCondition struct {
+	CloudPrivateIPConfigSpecItem `json:",inline" protobuf:"bytes,1,opt,name=cloudPrivateIPConfigSpecItem"`
+	// status is the status of the condition.
+	// +kubebuilder:validation:Enum="ERROR";"PENDING";"SUCCESS"
+	// +kubebuilder:validation:Required
+	// +required
+	Status ConditionStatus `json:"status" protobuf:"bytes,3,opt,name=status,casttype=ConditionStatus"`
```

Comment:
There are standard kubernetes conventions for Conditions (https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) and I think we should conform to those. Even if that's more complicated than what we actually need here, it's good to be consistent with other APIs (and may allow us to benefit from API-agnostic tooling later, etc)

Reply:
OK, so you're saying we change to something along the lines of:

```
apiVersion: cloud.network.openshift.io/v1
kind: CloudPrivateIPConfig
metadata:
  name: cluster
spec:
  items:
  - node: nodeX
    egressIP: 192.168.126.11
  - node: nodeY
    egressIP: 192.168.126.102
  - node: nodeZ
    egressIP: 192.168.126.15
status:
  items:
  - node: nodeX
    egressIP: 192.168.126.11
    message: Assignment successful
    reason: AssignmentReady
    status: "True"
    type: Ready
  - node: nodeY
    egressIP: 192.168.126.102
    message: Some reason why this has failed
    reason: AssignmentNotReady
    status: "False"
    type: Ready
  - node: nodeZ
    egressIP: 192.168.126.15
    message: "Assignment in progress"
    reason: AssignmentPending
    status: "False"
    type: Ready
```

Reply:
hm... now that you write that out, that's clearly wrong too.
the array should only have a single condition of each `type`... so either we'd need a conditions array per item, or we'd want to flip it inside out like:

```
status:
  conditions:
    - type: Ready
      status: True
      message: Assignment successful
      reason: AssignmentReady
      items:
        - node: nodeX
          egressIP: 192.168.126.11
        - ...
    - type: Failed
      ...
      items:
        - ...
    - type: Pending
      ...
      items:
        - ...
```

but I don't know if that makes sense either. Can one of the @openshift/api-reviewers comment?

(If we end up going back to how you had it before, the enum values should just be initial caps, not all caps.)